runtime/src/iree/builtins/ukernel/tools/e2e_matmul_benchmark.c

```diff
@@ -0,0 +1,366 @@
+      .in_size0 = params->M,
+      .in_size1 = params->K,
+      .out_size0 = M1,
+      .out_size1 = K1,
+      .out_size2 = M0,
+      .out_size3 = K0,
+      .in_stride0 = params->K,
+      .out_stride0 = mmt4d_params.lhs_stride,
+      .padding_value = padding_value_buf,
+  };
+
+  iree_uk_pack_params_t pack_rhs_params = {
+      .type = iree_uk_tie_2_types(rhs_type, rhs_type),
+      .cpu_data = cpu_data,
+      .flags =
+          IREE_UK_FLAG_PACK_TRANSPOSE_INNER | IREE_UK_FLAG_PACK_TRANSPOSE_OUTER,
+      .in_size0 = params->K,
+      .in_size1 = params->N,
+      .out_size0 = N1,
+      .out_size1 = K1,
+      .out_size2 = N0,
+      .out_size3 = K0,
+      .in_stride0 = params->N,
+      .out_stride0 = mmt4d_params.rhs_stride,
+      .padding_value = padding_value_buf,
+  };
+
+  iree_uk_pack_params_t pack_out_params = {
+      .type = iree_uk_tie_2_types(out_type, out_type),
+      .cpu_data = cpu_data,
+      .flags = 0,
+      .in_size0 = params->M,
+      .in_size1 = params->N,
+      .out_size0 = M1,
+      .out_size1 = N1,
+      .out_size2 = M0,
+      .out_size3 = N0,
+      .in_stride0 = params->N,
+      .out_stride0 = mmt4d_params.out_stride,
+      .padding_value = padding_value_buf,
+  };
+
+  iree_uk_unpack_params_t unpack_out_params = {
+      .type = iree_uk_tie_2_types(out_type, out_type),
+      .cpu_data = cpu_data,
+      .flags = 0,
+      .out_size0 = params->M,
+      .out_size1 = params->N,
+      .in_size0 = M1,
+      .in_size1 = N1,
+      .in_size2 = M0,
+      .in_size3 = N0,
+      .out_stride0 = params->N,
+      .in_stride0 = mmt4d_params.out_stride,
+  };
+
+  iree_uk_ssize_t rowmajor_lhs_buffer_size =
+      iree_uk_2d_buffer_length(lhs_type, params->M, params->K);
+  iree_uk_ssize_t rowmajor_rhs_buffer_size =
+      iree_uk_2d_buffer_length(rhs_type, params->K, params->N);
+  iree_uk_ssize_t rowmajor_out_buffer_size =
+      iree_uk_2d_buffer_length(out_type, params->M, params->N);
+  iree_uk_ssize_t packed_lhs_buffer_size =
+      iree_uk_2d_buffer_length(lhs_type, M1, mmt4d_params.lhs_stride);
+  iree_uk_ssize_t packed_rhs_buffer_size =
+      iree_uk_2d_buffer_length(rhs_type, N1, mmt4d_params.rhs_stride);
+  iree_uk_ssize_t packed_out_buffer_size =
+      iree_uk_2d_buffer_length(out_type, M1, mmt4d_params.out_stride);
+  void* rowmajor_lhs_buffer = malloc(rowmajor_lhs_buffer_size);
+  void* rowmajor_rhs_buffer = malloc(rowmajor_rhs_buffer_size);
+  void* rowmajor_init_out_buffer = malloc(rowmajor_out_buffer_size);
+  void* rowmajor_out_buffer = malloc(rowmajor_out_buffer_size);
+  void* packed_lhs_buffer = malloc(packed_lhs_buffer_size);
+  void* packed_rhs_buffer = malloc(packed_rhs_buffer_size);
+  void* packed_out_buffer = malloc(packed_out_buffer_size);
+  iree_uk_random_engine_t* engine = iree_uk_benchmark_random_engine(user_data);
+  // It's just about plausible that on some platform, for some number type,
+  // performance might be different on zero buffers vs random buffers. But it
```

Comment:
My favorite case was when someone's results was due to NANs having different performance on different devices and they happened to initialize with NANs :)

Reply:
Haha, yes, non-finite float values are heavily associated with performance impacts. Here though this is referring only to whether we can benchmark with zeros or with small finite values. That typically doesn't make an impact, but it's still possible that it would on some target.